### PR TITLE
BUGFIX: RAIL-4758 Can't save a dashboard with D2U to a hyperlink attribute

### PR DESCRIPTION
**Pull request by hackerstanislav**
In new KD edit mode, place an insight has at least 1 metric and 1 hyperlink attribute to KD. Try to set up D2U from a metric to the hyperlink attribute (don’t use custom url dialog), press save the dashboard

 + use factories where is it possible

JIRA: RAIL-4758

---

Supported PR commands:

| Command                  | Description             |
| ------------------------ | ----------------------- |
| `ok to test`             | Re-run standard checks  |
| `extended test`          | BackstopJS tests        |
| `extended check sonar`   | SonarQube tests         |
| `extended check cypress` | Cypress E2E tests       |
| `extended check plugins` | Dashboard plugins tests |

---

# PR Checklist

-   [ ] commit messages adhere to the [commit message guidelines](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#what-should-the-commits-look-like)
-   [ ] review was done by a Code owner [if necessary](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-tell-if-my-pull-request-needs-approval-by-a-code-owner) (if you think it is not necessary, explain the reasoning in the description or in a comment)
-   [ ] `check` passes
-   [ ] `check-extended` passes
-   [ ] `check-extended-cypress` passes
-   [ ] `rush change` [was run if applicable](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-describe-my-changes-for-the-changelog)

### CODE REVIEW

File: libs/sdk-model/src/objRef/factory.ts [hunk 1]
```diff
@@ -34,3 +34,14 @@ export function uriRef(uri: Uri): UriRef {
...
 }
+
+/**
+ * Creates an ObjRef from a id and uri
+ * @param uri - URI to use
+ * @param identifier - identifier to use
+ * @returns obj ref reference
+ * @public
+ */
+export function objRef(uri: string, identifier: string): ObjRef {
```
Comment (no23reason) [inline libs/sdk-model/src/objRef/factory.ts:45]: I am not sure we want to promote this "pattern". At least we should make this internal or something.
@kandl WDYT?

File: libs/sdk-backend-bear/src/convertors/toBackend/DashboardConverter.ts [hunk 2]
```diff
@@ -75,13 +74,13 @@ import omitBy from "lodash/omitBy";
 import { serializeProperties } from "../fromBackend/PropertiesConverter";
 import { assertNoNulls } from "./utils";
 
-const refToUri = (ref: ObjRef) => {
+const refToUri = (ref: unknown) => {
```
Comment (no23reason) [inline libs/sdk-backend-bear/src/convertors/toBackend/DashboardConverter.ts:77]: Why is this change needed?

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/10908/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/12995/ : SUCCESS in 19m 54s

File: libs/sdk-model/src/objRef/factory.ts (hunk 1 above)
Comment (hackerstanislav) [reply libs/sdk-model/src/objRef/factory.ts]: Moved to dashboard package.

File: libs/sdk-backend-bear/src/convertors/toBackend/DashboardConverter.ts (hunk 2 above)
Comment (hackerstanislav) [reply libs/sdk-backend-bear/src/convertors/toBackend/DashboardConverter.ts]: Reverted, forgotten change

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/10915/ : SUCCESS in 6s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/13003/ : SUCCESS in 19m 47s

**Review by no23reason — APPROVED** [review]
LGTM

File: libs/sdk-backend-bear/src/convertors/fromBackend/ObjRefConverter.ts [hunk 3]
```diff
@@ -10,12 +10,12 @@ import { isLocalIdRef, isUriRef, ObjectType, ObjRef, ObjRefInScope } from "@good
  * @param defaultType - type to use it the ref has none specified
  * @internal
  */
-export function fromBearRef(ref: ObjRef, defaultType?: ObjectType): ObjRef {
-    if (isUriRef(ref)) {
-        return ref;
+export function fromBearRef(ref: GdcVisualizationObject.ObjQualifier, defaultType?: ObjectType): ObjRef {
+    if (GdcExecuteAFM.isObjectUriQualifier(ref)) {
```
Comment (ivanmjartan) [inline libs/sdk-backend-bear/src/convertors/fromBackend/ObjRefConverter.ts:14]: Why you need use this guard ? GdcExecuteAFM.isObjectUriQualifier ? isUriRef is not usable ?

File: libs/sdk-backend-bear/src/convertors/fromBackend/ObjRefConverter.ts [hunk 4]
```diff
@@ -10,12 +10,12 @@ import { isLocalIdRef, isUriRef, ObjectType, ObjRef, ObjRefInScope } from "@good
...
+        return uriRef(ref.uri);
     }
 
-    return { identifier: ref.identifier, type: ref.type ?? defaultType };
+    return idRef(ref.identifier, defaultType);
```
Comment (ivanmjartan) [inline libs/sdk-backend-bear/src/convertors/fromBackend/ObjRefConverter.ts:18]: Its ok this conversion when defaultType is missing and ref.type present ?

File: libs/sdk-backend-bear/src/convertors/fromBackend/ObjRefConverter.ts (hunk 3 above)
Comment (hackerstanislav) [reply libs/sdk-backend-bear/src/convertors/fromBackend/ObjRefConverter.ts:14]: Because of typings, that object is same, but check wrong type

File: libs/sdk-backend-bear/src/convertors/fromBackend/ObjRefConverter.ts (hunk 4 above)
Comment (hackerstanislav) [reply libs/sdk-backend-bear/src/convertors/fromBackend/ObjRefConverter.ts:18]: Basically we can do this params as required, because all places have this apram filled

File: libs/sdk-backend-bear/src/convertors/fromBackend/ObjRefConverter.ts (hunk 3 above)
Comment (ivanmjartan) [reply libs/sdk-backend-bear/src/convertors/fromBackend/ObjRefConverter.ts:14]: Ok thanks

**Comment by yenkins** [comment]
### Sonar scan result


More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-ui-sdk-gate-PR3345

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

**Comment by gdgate** [comment]
Build in pipeline gate aborted.

**Review by ivanmjartan — APPROVED** [review]

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by yenkins** [comment]
### Sonar scan result


More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-ui-sdk-gate-PR3345

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-fossa-scanning-zuul/215/ : FAILURE in 2m 31s (non-voting)
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-merge-compliance-check/3481/ : SUCCESS in 40s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-sonarqube-scan-zuul-gate/2330/ : SUCCESS in 27m 44s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/13008/ : SUCCESS in 34m 20s
